pkg/operator/controller/dnsnameresolver/controller.go

```diff
@@ -0,0 +1,217 @@
+// objects. The controller re-resolves the DNS names which get added to the
+// status of the DNSNameResolver objects. It also removes IP addresses of DNS
+// names, whose TTLs have expired, from the staus. This is an unmanaged
+// controller, which means that the manager does not start it.
+func NewUnmanaged(mgr manager.Manager, config Config) (controller.Controller, []cache.Cache, error) {
+	// Create a new cache for tracking the DNSNameResolver resources in
+	// the DNSNameResolverNamespace.
+	dnsNameResolverCache, err := cache.New(mgr.GetConfig(), cache.Options{
+		Scheme: mgr.GetScheme(),
+		DefaultNamespaces: map[string]cache.Config{
+			config.DNSNameResolverNamespace: {},
+		},
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Create a new cache to track the EndpointSlices corresponding to the
+	// CoreDNS pods.
+	corednsEndpointsSliceCache, err := cache.New(mgr.GetConfig(), cache.Options{
+		Scheme: mgr.GetScheme(),
+		DefaultNamespaces: map[string]cache.Config{
+			config.OperandNamespace: {},
+		},
+		DefaultLabelSelector: labels.SelectorFromSet(labels.Set{
+			discoveryv1.LabelServiceName: config.ServiceName,
+		}),
+	})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Initialize the reconciler.
+	reconciler := &reconciler{
+		config:               config,
+		dnsNameResolverCache: dnsNameResolverCache,
+		client:               mgr.GetClient(),
+		res:                  NewResolver(corednsEndpointsSliceCache, config.ServiceName),
+	}
+
+	// Create an unmanaged controller using the reconciler.
+	c, err := controller.NewUnmanaged(controllerName, mgr, controller.Options{Reconciler: reconciler})
+	if err != nil {
+		return nil, nil, err
+	}
+
+	// Watch for the DNSNameResolver resources using dnsNameResolverCache.
+	if err := c.Watch(source.Kind(dnsNameResolverCache, &networkv1alpha1.DNSNameResolver{}), &handler.EnqueueRequestForObject{}); err != nil {
+		return nil, nil, err
+	}
+
+	// Watch for the CoreDNS pod EndpointSlices using corednsEndpointsSliceCache.
+	if err := c.Watch(source.Kind(corednsEndpointsSliceCache, &discoveryv1.EndpointSlice{}), handler.EnqueueRequestsFromMapFunc(skipReconcileForEndpointSlice)); err != nil {
+		return nil, nil, err
+	}
+
+	return c, []cache.Cache{dnsNameResolverCache, corednsEndpointsSliceCache}, nil
+}
+
+// skipReconcileForEndpointSlice just skips the reconcile if any event related to the CoreDNS pod
+// endpointslices is received.
+func skipReconcileForEndpointSlice(context.Context, client.Object) []reconcile.Request {
+	return nil
+}
+
+type Config struct {
+	OperandNamespace         string
+	ServiceName              string
+	DNSNameResolverNamespace string
+}
+
+// reconciler handles the actual DNSNameResolver reconciliation logic in response to events.
+type reconciler struct {
+	config Config
+
+	dnsNameResolverCache cache.Cache
+	client               client.Client
+	res                  *Resolver
+	startResolver        sync.Once
+}
```

Comment:
Can you clean this up and move the type struct declarations before the func declarations?

Reply:
Moved to openshift/coredns-ocp-dnsnameresolver#2